web-api/src/storage/elastic.rs

```diff
@@ -186,39 +192,94 @@ impl Client {
 
     pub(super) async fn insert_documents(
         &self,
-        documents: impl IntoIterator<
-            IntoIter = impl ExactSizeIterator<Item = &models::IngestedDocument>,
-        >,
+        documents: &[models::IngestedDocument],
     ) -> Result<Warning<DocumentId>, Error> {
-        let documents = documents.into_iter();
-        if documents.len() == 0 {
+        if documents.is_empty() {
             return Ok(Warning::default());
         }
 
-        let response = self
-            .bulk_request(documents.flat_map(|document| {
-                [
-                    serde_json::to_value(BulkInstruction::Index { id: &document.id })
-                        .map_err(Into::into),
-                    serde_json::to_value(IngestedDocument {
+        let pre_deletions = documents
+            .iter()
+            .filter_map(|document| (document.embeddings.len() > 1).then_some(&document.id))
+            .collect_vec();
+
+        self.delete_by_parents(pre_deletions).await?;
```

Comment:
:thinking:  We could make `delete_by_parents` non-blocking by setting both `wait_for_completion=false` and `conflicts=proceed`.

Due to how `delete_by_parents` uses the ES versioning system it would lead to only deleting documents which haven't been changed since the start of the delete operation. If they where changed since then (through the operation below) it would fail but due to `conflict=proceed` this would be fine, too. 

Through not sure if we should.

Reply:
> as we still have "refresh" in every es call, we should also keep this call waiting for completion. we can reconsider this once we deal with "refresh" again.

we did strip `refresh` already out from some calls and it also doesn't have anything to do directly with `refresh` as such call could always being before a refresh is done and complete after the refresh, i.e. refresh isn't really a synchronization primitive, it just makes some background updates run earlier and maybe less in the background AFIK